### PR TITLE
Issue #1375: Return a Robo\Result to set the proper error code.

### PR DESCRIPTION
**Pull request by geerlingguy**
[WIP] - See issue #1375.

### CODE REVIEW

**Comment by grasmash** [comment]
Still a WIP?

**Comment by geerlingguy** [comment]
@grasmash yeah, this still doesn't actually pass the error code all the way back through the blt command.

Something in BLT's custom Robo application seems to eat the error code. I was tracing it yesterday and couldn't figure out what. The `$errorCode` is `1` internally when the behat task is run... but by the time it gets back to the `Blt` robo app, it's `0` :/

It seems you're following the docs with how you use Robo as a framework, but I fear maybe the docs are missing something essential. Or maybe it's a bug in Robo itself.

Anyways, you can test this issue by doing the following:

  1. Add one failing Behat test.
  2. Run behat tests using `blt tests:behat`
  3. After you get the failure in stdout, run `echo $?` — it returns `0` even though Robo's output says `Exit Code: 1`.

File: src/Robo/Commands/Tests/BehatCommand.php [hunk 1]
```diff
@@ -198,8 +200,9 @@ protected function executeBehatTests() {
       $result = $task->run();
 
       if (!$result->wasSuccessful()) {
-        throw new \Exception("Behat tests failed");
```
Comment (grasmash) [inline src/Robo/Commands/Tests/BehatCommand.php]: I'm surprised that throwing an exception doesn't stop the execution?

Comment (grasmash) [inline src/Robo/Commands/Tests/BehatCommand.php]: Oh. Maybe it does, but it's still a 0 exit code.

**Comment by grasmash** [comment]
@greg-1-anderson any thoughts on this?

**Comment by grasmash** [comment]
@geerlingguy I think this is simply because executeBehatTests() returns the error, but behat() doesn't return anything to Robo.

The intention with throwing an exception was to bypass the return chain and just exit immediately.

**Comment by grasmash** [comment]
Wait throwing an exception _does_ create an exit code of 1.

**Comment by grasmash** [comment]
I think this may resolve it, but I can't reproduce the issue in the first place: #1381

**Comment by geerlingguy** [comment]
@grasmash - I had tracked it down to here locally:

```php
        try {
            $statusCode = $app->run($input, $output);
        } catch (TaskExitException $e) {
            // HERE - even if I throw a `TaskExitException` this never gets hit.
            $statusCode = $e->getCode() ?: 1;
        }
        return $statusCode;
```

That's from robo's `Runner.php` class.

Basically, I could throw a `TaskExitException` inside a Task, but it seems something in Blt ate it up, and no exception made it back to that `run()` method.

**Comment by geerlingguy** [comment]
Steps to reproduce in Canary (private repo) on my branch:

  1. Build a docker container for the project: `container_id=blttest ./tests/scaffold/build-docker-image.sh`
  1. Log into the container: `docker exec -it blttest /bin/bash`
  1. Set the $PATH: `export PATH=$PATH:/var/www/drupal/vendor/bin`
  1. cd into the project dir: `cd /var/www/drupal/docroot`
  1. (Make sure one of the Behat tests in the `tests/behat/features` dir fails—e.g. change a condition or something).
  1. Run ci:setup: `blt ci:setup -Dcreate_alias=false`
  1. Run Behat tests: `blt tests:all --define behat.run-server=true --yes -v`
  1. Observe the failure, then run `echo $?` to see if the exit code is 1 or 0.

For example:

```
  Scenario: Load a page without Javascript # features/Examples.feature:13
    Given I am on "/"                      # Drupal\DrupalExtension\Context\MinkContext::visit()
    Then I should see the text "Log on"    # Drupal\DrupalExtension\Context\MinkContext::assertTextVisible()
      The text "Log on" was not found anywhere in the text of the current page. (Behat\Mink\Exception\ResponseTextException)

--- Failed scenarios:

    features/Examples.feature:13

3 scenarios (2 passed, 1 failed)
12 steps (11 passed, 1 failed)
0m11.53s (72.83Mb)
[Testing\Behat]  Exit code 1 
[error]  Behat tests failed 

root@d63a8eec3941:/var/www/drupal# echo $?
0
```

I've only reproduced _inside_ the docker container for now; I will set up a full VM in a little while and test there next.

**Comment by greg-1-anderson** [comment]
Not sure why the exceptions are not passing through here.

Any chance you might want to re-frame these functions as chained tasks?
```
$this
  ->taskSomething()
     ->somethingParam($a)
  ->taskSomethingElse()
     ->someOtherParam($b)
  ->run();
```
All of the tasks in the chain run, unless one of them errors, in which case the tasks that follow do not run. You should then always get a `RoboResult` at the end.

You can write routines that just return a collection of tasks (as above, but without the `run()`; these can be added to collections.

**Comment by grasmash** [comment]
Yes I have the intention of converting these to task collections. Though, for the time being, the Exception should produce the correct exit code.

**Comment by geerlingguy** [comment]
@grasmash - Also when running inside Drupal VM:

```
  Scenario: Load a page without Javascript # features/Examples.feature:13
    Given I am on "/"                      # Drupal\DrupalExtension\Context\MinkContext::visit()
    Then I should see the text "Log on"    # Drupal\DrupalExtension\Context\MinkContext::assertTextVisible()
      The text "Log on" was not found anywhere in the text of the current page. (Behat\Mink\Exception\ResponseTextException)

--- Failed scenarios:

    features/Examples.feature:13

3 scenarios (2 passed, 1 failed)
12 steps (11 passed, 1 failed)
0m6.47s (84.62Mb)
[Testing\Behat]  Exit code 1 
[error]  Behat tests failed 

vagrant@local:/var/www/canary$ echo $?
0
```

**Comment by geerlingguy** [comment]
I'm going to nuke my entire repo locally and re-clone to verify that there are no other changes.

**Comment by geerlingguy** [comment]
@grasmash it's occurring on Travis CI as well (no VM or Docker involved):

```
  Scenario: Load a page without Javascript     # features/Examples.feature:13
    Given I am on "/"                          # Drupal\DrupalExtension\Context\MinkContext::visit()
    Then I should see the text "Log not found" # Drupal\DrupalExtension\Context\MinkContext::assertTextVisible()
      The text "Log not found" was not found anywhere in the text of the current page. (Behat\Mink\Exception\ResponseTextException)
--- Failed scenarios:
    features/Examples.feature:13
3 scenarios (2 passed, 1 failed)
12 steps (11 passed, 1 failed)
0m4.00s (78.89Mb)
[Testing\Behat]  Exit code 1 
[error]  Behat tests failed 

The command "source $BLT_DIR/scripts/travis/run_tests" exited with 0.
```

**Comment by geerlingguy** [comment]
Testing #1377 now, locally inside Drupal VM. If it works, I'll also try it inside the Docker container.

**Comment by geerlingguy** [comment]
@grasmash It looks like #1377 is the fix! See:

```
3 scenarios (2 passed, 1 failed)
12 steps (11 passed, 1 failed)
0m8.54s (82.05Mb)
[Testing\Behat]  Exit code 1 
[error]  Behat tests failed 

vagrant@local:/var/www/canary$ echo $?
1
```

Before, that was a `0`. Applied the patch and now it's a `1`.

**Pull request closed without merging**